### ISSUE TITLE
Swtpm_setup error when I create a certificate

### ISSUE THREAD

**Issue by ferdinan4** [stefanberger/swtpm#184]
Hi Stefan!,

When I am trying to create certificates for a tpm2.0 I got an error. I tried to follow your wiki, but it doesn't work for me.

after launch the command : **sudo swtpm_setup --tpmstate /tmp/myvtpm2 --create-ek-cert --create-platform-cert --allow-signing --tpm2**

Starting vTPM manufacturing as root:root @ Thu 28 Nov 2019 09:33:39 GMT

TPM is listening on TCP port 53659.
Error: TPM2_Stirrandom() failed
Error:        expected:  80 01 00 00 00 0a 00 00 00 00
Error:        received:  80 01 00 00 00 0a 00 00 01 d5
Error: TPM2_ChangeEPS() failed
Error:        expected:  80 02 00 00 00 13 00 00 00 00 00 00 00 00 00 00 01 00 00
Error:        received:  80 01 00 00 00 0a 00 00 09 9a
Error: tpm2_create_ek failed
Error: An error occurred. Authoring the TPM state failed.
Ending vTPM manufacturing @ Thu 28 Nov 2019 09:33:39 GMT

any idea?

Thanks in advance

**Comment by stefanberger** [comment]
I don't know what could be wrong with your setup. The test cases for swtpm cover certificate creation, so does `make check`  succeed, in case you have the sources installed? You should be able to run all these test case with a non-root user.

```
# ./tests/test_tpm2_swtpm_setup_create_cert
TPM is listening on TCP port 41437.
Test 1: OK
TPM is listening on TCP port 55315.
Test 2: OK
# ./tests/test_tpm2_swtpm_cert
Test 1: OK
Test 2: OK
Test 3: OK
Test 4: OK
# ./tests/test_tpm2_swtpm_cert_ecc
Test 1: OK
Test 2: OK
Test 3: OK
Test 4: OK
```

**Comment by ferdinan4** [comment]
Hi Stefan, Sorry for the delay in my answer, I was trying to figure out where it is the problem: 

I **am doing some modification in the libtpms, in order to add new algorithms (Do you have any example of which files we have to modify to follow as a reference)**.

I guess at the beguinning that due these modifications, I have this problem:

The tests: 
[krilin4@localhost tests]$ ./test_tpm2_swtpm_cert_ecc 
Test 1: OK
Test 2: OK
Test 3: OK
Test 4: OK

[krilin4@localhost tests]$ ./test_tpm2_swtpm_cert
Test 1: OK
Test 2: OK
Test 3: OK
Test 4: OK

works well but, the problem it is in 
[krilin4@localhost swtpm-0.2.0]$ **./tests/test_tpm2_swtpm_setup_create_cert**
TPM is listening on TCP port 61439.
Error: Could not run ./tests/../src/swtpm_setup/swtpm_setup.
Logfile output:
Starting vTPM manufacturing as krilin4:krilin4 @ Wed 11 Dec 2019 15:44:43 GMT
Error: TPM2_Stirrandom() failed
Error:        expected:  80 01 00 00 00 0a 00 00 00 00
Error:        received:  80 01 00 00 00 0a 00 00 01 d5
Error: TPM2_ChangeEPS() failed
Error:        expected:  80 02 00 00 00 13 00 00 00 00 00 00 00 00 00 00 01 00 00
Error:        received:  80 01 00 00 00 0a 00 00 09 9a
Error: tpm2_create_ek failed
Error: An error occurred. Authoring the TPM state failed.
Ending vTPM manufacturing @ Wed 11 Dec 2019 15:44:43 GMT


**I found in another issue** related with this  problem, talking about the permissions of user tss, so after executing this command:

             **sudo chown tss: /tmp/myvtpm2/**

             **sudo swtpm_setup --tpm-state /tmp/myvtpm2 --create-platform-cert --create-ek-cert**

TPM is listening on TCP port 35275.
Successfully created EK.
TSS is listening on TCP port 46067.
Successfully took ownership of the TPM.
  Invoking: /usr/share/swtpm/swtpm-localca --type ek --ek 9ff971fb37bc085f8dffb6c5969130d28e375fd0d6b60d65bda6cfa520fe5bf54afbb93b7566f6e1f1545b4dc1a5f3ad8408e43ccd0f0a618eb2df7da65f706e1897cd6e262bd61b44836e2c82eb3d72b1f6cb68ed84cc93a114e53144438a886d5e01142fe9eaed29b432b72ad01d24fc06aff3c5aa2b495e377c21bbba63f2f5ed7b711e2d08f2ef2ea4a9827f88d9d0956e5e28faae1c730c3537236ac04de02496c1dc822b34535ea61f580a86cea76559bdb5500408a690092124990d26a91b39d436ddb96b467caddb4104f12cc7f199105c9a6642f2b7682a2c857825d8d47a1b3610aaabb3f8ed0b98e4fd49e57fa9992afd03656d092818617490c3 --dir /tmp/myvtpm2 --tpm-spec-family 1.2 --tpm-spec-level 2 --tpm-spec-revision 116 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:00740001 --configfile "/etc/swtpm-localca.conf" --optsfile "/etc/swtpm-localca.options"
swtpm-localca: Creating root CA and a local CA's signing key and issuer cert.
swtpm-localca: Successfully created EK certificate locally.
  Invoking: /usr/share/swtpm/swtpm-localca --type platform --ek 9ff971fb37bc085f8dffb6c5969130d28e375fd0d6b60d65bda6cfa520fe5bf54afbb93b7566f6e1f1545b4dc1a5f3ad8408e43ccd0f0a618eb2df7da65f706e1897cd6e262bd61b44836e2c82eb3d72b1f6cb68ed84cc93a114e53144438a886d5e01142fe9eaed29b432b72ad01d24fc06aff3c5aa2b495e377c21bbba63f2f5ed7b711e2d08f2ef2ea4a9827f88d9d0956e5e28faae1c730c3537236ac04de02496c1dc822b34535ea61f580a86cea76559bdb5500408a690092124990d26a91b39d436ddb96b467caddb4104f12cc7f199105c9a6642f2b7682a2c857825d8d47a1b3610aaabb3f8ed0b98e4fd49e57fa9992afd03656d092818617490c3 --dir /tmp/myvtpm2 --tpm-spec-family 1.2 --tpm-spec-level 2 --tpm-spec-revision 116 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:00740001 --configfile "/etc/swtpm-localca.conf" --optsfile "/etc/swtpm-localca.options"
swtpm-localca: Successfully created platform certificate locally.
Successfully created NVRAM area for EK certificate.
Successfully created NVRAM area for platform certificate.
Successfully gave up ownership of the TPM.
Successfully enabled and activated the TPM

Successfully authored TPM state.

Looks as it is working properly.

Kind regards,

Fernando

**Comment by stefanberger** [comment]
What algorithm are you trying to add?

**Comment by stefanberger** [comment]
As for this error here:

Error: TPM2_Stirrandom() failed
Error: expected: 80 01 00 00 00 0a 00 00 00 00
Error: received: 80 01 00 00 00 0a 00 00 01 d5

The error code 0x1d5 means 'structure is the wrong size Parameter number 1'. Did you change something on the swtpm_setup side?

**Comment by ferdinan4** [comment]
No, Just some modification in libtpms. :S

Do you know how can I trace the error?

Thanks

**Comment by ferdinan4** [comment]
I am trying to add Kyber

**Comment by stefanberger** [comment]
You could either look for a simpler client tool to cause the issue or you would have to go into swtpm_setup.sh{.in} and add a 'sleep 20' or something like that into the script to give you time to determine the process id of swtpm and then hook gdb via `gbd pid <pid>` onto swtpm and then set a breakpoint at `TPM2_Process` or at '_rpc__Send_Command' to then single step through the code. 
Another strategy would be to disable some of the code you added to libtpms to see which changes are breaking the client tool.

**Comment by stefanberger** [comment]
As for Kyper I would suggest to go through the TPM working group to get this [spec'ed and] accepted. Prototyping with libtpms is fine, of course, but TCG needs to accept it and possibly extend specs and add it to the algorithm registry.

**Comment by stefanberger** [comment]
I don't think I can help with this issue and since this error occurs only after you made modifications to libtpms, can you close it?

**Comment by stefanberger** [comment]
This is what the output of the compiled version should be when you are in the swtpm git checkout directory:
```
./src/swtpm_cert/swtpm_cert --help | grep days -A1
--days <number>           : Number of days the cert is valid;
                            -1 for no expiration
```

The installed version should have the same:

```
swtpm_cert --help | grep days -A1
--days <number>           : Number of days the cert is valid;
                            -1 for no expiration
```

Maybe you don't have gnutls-devel / gnutls-dev package installed on your system and swtpm_cert didn't get built and installed. You need this package and then `./autogen --prefix=/usr` again, build, and then install.

**Comment by stefanberger** [comment]
The name of the package on Ubuntu is `gnutls-dev`.
You have to see this configure/autogen.sh output to have `swtpm_cert` built:

```
[...]
with_gnutls     :   yes  (no = swtpm_cert will NOT be built)
[...]
```

**Comment by OliverLeitner** [comment]
also...
which swtpm_cert 
/usr/local/bin/swtpm_cert

**Comment by stefanberger** [comment]
If you have an older version on your system you will have to install the newer version on top of the older one or remove the older one first like this unless you can remove it as a package:

```
./configure --prefix=/usr/local
sudo make uninstall
./configure --prefix=/usr
make -j32
sudo make install
```
